### PR TITLE
Tidy up basic xthinblock checking and use atomics for tracking requests

### PR DESCRIPTION
**Pull request by ptschip**

### CODE REVIEW

File: src/blockrelay/graphene.cpp [hunk 1]
```diff
@@ -234,14 +234,14 @@ bool CRequestGrapheneBlockTx::HandleMessage(CDataStream &vRecv, CNode *pfrom)
...
         if (pfrom->nGetGrapheneBlockTxLastTime <= 0)
             pfrom->nGetGrapheneBlockTxLastTime = GetTime();
         uint64_t nNow = GetTime();
-        pfrom->nGetGrapheneBlockTxCount *=
+        pfrom->nGetGrapheneBlockTxCount =
+            pfrom->nGetGrapheneBlockTxCount *
             std::pow(1.0 - 1.0 / 600.0, (double)(nNow - pfrom->nGetGrapheneBlockTxLastTime));
         pfrom->nGetGrapheneBlockTxLastTime = nNow;
-        pfrom->nGetGrapheneBlockTxCount += 1;
+        pfrom->nGetGrapheneBlockTxCount = pfrom->nGetGrapheneBlockTxCount + 1;
```
Comment (gandrewstone) [inline src/blockrelay/graphene.cpp]: Shouldn't you use the ++ operator here (and 3 places below) to ensure that pfrom->nGetGrapheneBlockTxCount is atomically incremented?

File: src/blockrelay/graphene.cpp [hunk 2]
```diff
@@ -234,14 +234,14 @@ bool CRequestGrapheneBlockTx::HandleMessage(CDataStream &vRecv, CNode *pfrom)
     // Check for Misbehaving and DOS
     // If they make more than 20 requests in 10 minutes then disconnect them
     {
-        LOCK(cs_vNodes);
         if (pfrom->nGetGrapheneBlockTxLastTime <= 0)
             pfrom->nGetGrapheneBlockTxLastTime = GetTime();
         uint64_t nNow = GetTime();
-        pfrom->nGetGrapheneBlockTxCount *=
+        pfrom->nGetGrapheneBlockTxCount =
```
Comment (gandrewstone) [inline src/blockrelay/graphene.cpp]: even with atomics, this may not work because: 
In thread 1 pfrom->nGetGrapheneBlockTxCount is read and put into a temporary while the rest of the expression is calculated.  Then thread 2 enters this same block and modifies pfrom->nGetGrapheneBlockTxCount.  Then thread 1 continues and overwrites pfrom->nGetGrapheneBlockTxCount with its calculated value.  So thread 2's modification got lost.

To do this atomically we need to use compare_exchange_xxx.
```
pfrom->nGetGrapheneBlockTxCount =
            pfrom->nGetGrapheneBlockTxCount * std::pow(1.0 - 1.0 / 600.0, (double)(nNow - pfrom->nGetGrapheneBlockTxLastTime));
```
becomes:
```
double tmp = pfrom->nGetGrapheneBlockTxCount;
while (!pfrom->nGetGrapheneBlockTxCount.compare_exchange_weak(tmp, tmp * std::pow(1.0 - 1.0 / 600.0, (double)(nNow - pfrom->nGetGrapheneBlockTxLastTime))));
```
basically compare_exchange_weak says atomically compare pfrom->nGetGrapheneBlockTxCount to "tmp" and assign the 2nd argument (tmp * std::pow(1.0 - 1.0 / 600.0, (double)(nNow - pfrom->nGetGrapheneBlockTxLastTime) ONLY if its equal.  If its not equal, load "tmp" with the new value of pfrom->nGetGrapheneBlockTxCount and try again.

This needs to be done in all places where *= is changed.

File: src/blockrelay/graphene.cpp (hunk 1 above)
Comment (ptschip) [reply src/blockrelay/graphene.cpp]: that doesn't work with <double> in C++11 for std::atomic<double>...I'll just add the +1 when I do the calculation above, no need to separately increment.

File: src/blockrelay/graphene.cpp (hunk 2 above)
Comment (ptschip) [reply src/blockrelay/graphene.cpp]: done...
